### DISCUSSION TITLE
Adapter Maintainers: Upgrading to dbt-core `1.6.0`

### DISCUSSION THREAD

**Discussion by dataders** [dbt-labs/dbt-core#7958] [Community]
## Overview 

This discussion is for communicating to adapter maintainers the scope of work needed to make use of the changes in 1.6.0. If you have questions and concerns, please ask them here for posterity.

Please consider this a living document between now and the date of final release. If there's something missing, please comment below!

### release timeline

The below table gives the milestones between up to and including the final release. It will be updated with each subsequent release

| **state**          | **date** | **stage**    | **version**    | **release**                                        | Diff to `1.5.0`                                                                     |
| ------------------ | -------- | ------------ | -------------- | -------------------------------------------------- | ----------------------------------------------------------------------------------- |
| :white_check_mark: | Jun 23   | beta         | `1.6.0b6`      | [PyPI](https://pypi.org/project/dbt-core/1.6.0b6/) | [compare `1.6.0b6`](https://github.com/dbt-labs/dbt-core/compare/v1.5.0...v1.6.0b6) |
| :white_check_mark:     | Jul 13   | release cut  | `1.6.0rc1`     |                                                    |                                                                                     |
| :white_check_mark:     | Jul 27   | final        | `1.6.0`        |                                                    |                                                                                     |
| :white_check_mark:     | Jul 31   | in dbt Cloud | `1.6 (latest)` |                                                    |                                                                                     |

### prior maintainer upgrade versions

- #7213
- #6624
- #6011
- #5468

## Changes

as an adapter maintainer, there's three options for new features in dbt Core:

1. support the change
2. "stub" the feature as in add in code to let users know that the feature is "not supported"
3. do nothing. if an adapter maintainer does nothing, then, if an end user tries to use that feature, the default implementation of dbt-core will be made which will result in SQL being generated and sent to the underlying data platform, which will likely make the platform throw an error.

### Example diffs from dbt Labs-owned adapters

Below gives the changes as of the latest beta release (as of June 26). The canonical way to check the diff of a minor version is `compare/v1.4.0...1.5.latest`, but the `1.5.latest` branch is not created until the `rc` is released.

- <https://github.com/dbt-labs/dbt-redshift/compare/v1.5.0...v1.6.0b4>
- <https://github.com/dbt-labs/dbt-bigquery/compare/v1.5.0...v1.6.0b4>
- <https://github.com/dbt-labs/dbt-snowflake/compare/v1.5.0...v1.6.0b3>
- <https://github.com/dbt-labs/dbt-spark/compare/v1.5.0...v1.6.0b3>

### Areas to consider

note:  :construction: means that this guide is not yet complete and "BtS" is short for Behind the Scenes, ie not a user-facing change

<details>

<summary> [FEATURE] Materialized Views</summary>

#### Context

original issue: https://github.com/dbt-labs/dbt-core/issues/6911

#### How to support

a more comprehensive guide is still forthcoming, but for now, please refer to the following PRs to learn more

relevant PRs:
- https://github.com/dbt-labs/dbt-core/pull/7334/
- https://github.com/dbt-labs/dbt-redshift/pull/387
- :construction: https://github.com/dbt-labs/dbt-snowflake/pull/659/
- :construction: https://github.com/dbt-labs/dbt-bigquery/issues/672

Of particular interested would are:
1. the default (global) implementation for materialized views ([`core/dbt/include/global_project/macros/materializations/models/materialized_view/materialized_view.sql`](https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/include/global_project/macros/materializations/models/materialized_view/materialized_view.sql))
2. [relation_configs/README.md](https://github.com/dbt-labs/dbt-core/pull/7239/files#diff-0f50b6142889a932591ab8dd774fac2a0dc075f2d7dfb8fbe50bb12fd02f1d64) which describes an extra config set related to MVs that likely will be embraced for all relation configuration in future minor versions
3. how postgres tweaks/overrides specific macros corresponding to the default/global implementation ([`plugins/postgres/dbt/include/postgres/macros/materializations/materialized_view.sql`](https://github.com/dbt-labs/dbt-core/blob/main/plugins/postgres/dbt/include/postgres/macros/materializations/materialized_view.sql)))
   1. `postgres__get_alter_materialized_view_as_sql`
   2. `postgres__get_create_materialized_view_as_sql`
   3. `postgres__get_replace_materialized_view_as_sql`
   4. `postgres__get_materialized_view_configuration_changes`
   5. `postgres__refresh_materialized_view`
   6. `postgres__update_indexes_on_materialized_view`
   7. `postgres__describe_materialized_view`
4. how dbt-snowflake implements dynamic tables (see https://github.com/dbt-labs/dbt-snowflake/pull/659/)
   1. [`materialization: dynamic_table`](https://github.com/dbt-labs/dbt-snowflake/blob/aa7bfd757de10d4beb0e55f729791d815107cfe8/dbt/include/snowflake/macros/materializations/dynamic_table/materialization.sql)
   2. `snowflake__create_table_as` (add a `is_dynamic` conditional)
   3. `snowflake__drop_relation_sql` (add a `is_dynamic` conditional)
   4. `snowflake__alter_dynamic_table_sql`
   5. `snowflake__create_dynamic_table_sql`
   6. `snowflake__describe_dynamic_table`
   7. `snowflake__drop_dynamic_table_sql`
   8. `snowflake__refresh_dynamic_table_sql`
   9. `snowflake__replace_dynamic_table_sql`
   10. `snowflake__alter_dynamic_table_sql_with_on_configuration_change_option`
   11. `dynamic_table_execute_no_op`
   12. `dynamic_table_execute_build_sql`

#### How to stub elegantly

to be completed

#### What if you do nothing

The default MV DDL statements will be sent to your engine, that will react accordingly - unless you already support MVs, in which case your implementation will superseed this one

</details>

<details>

<summary>[BtS] Drop support for Py 3.7</summary>

#### Context 

see #7082. As of June 2023, Python 3.7 is now “End of Life” (EOL)

#### How to (remove) support 

modify the `python_requires` specifier in your packages [`setup.py`](http://setup.py) as well as any other mentions of `3.7` to use `3.8` as the minimum version. Also give yourself the gift of not testing against 3.7 moving forward.

#### What if you do nothing 

You'll likely get security bots flagging vulnerability issues, and users may encounter strange bugs/errors for which there will be no official fix from the Python Software Foundation

</details>

<details>

<summary> [FEATURE] `dbt clone`</summary>

#### Context 

`dbt clone` ([docs page](https://docs.getdbt.com/reference/commands/clone))

#### How to support 

If your data platform supports the capability to clone, then there are two macros to override:

- `can_clone_table()`, and
- `create_or_replace_clone()`

See below for the versions introduced to the BigQuery adapter via [dbt-bigquery#784](https://github.com/dbt-labs/dbt-bigquery/pull/784).

```sql
{% macro bigquery__can_clone_table() %}
    {{ return(True) }}
{% endmacro %}

{% macro bigquery__create_or_replace_clone(this_relation, defer_relation) %}
    create or replace
      table {{ this_relation }}
      clone {{ defer_relation }}
{% endmacro %}
```

#### What if you do nothing 

tbc

</details>

<details>

<summary>[BtS] revamp of `dbt debug`</summary>

#### Context 

See [dbt-core#7104](https://github.com/dbt-labs/dbt-core/issues/7104)

#### How to support 

There is a new Adapter method, `.debug_query()`, whose default value is `select 1 as id`. If this does not work on your supported data platform, you may override it.

Also, an existing test was modified to test the new command-line flag functionality.

`TestDebugPostgres` ([sauce](https://github.com/dbt-labs/dbt-core/blob/adc4dbc4d6a2e4423a0e5159acc0c2f5d94f060f/tests/adapter/dbt/tests/adapter/dbt_debug/test_dbt_debug.py#L49C7-L82))

#### What if you do nothing 

no end-user impact

</details>

<details>

<summary> [BtS] new arg for `adapter.execute()`</summary>

#### Context 

To more fully support `dbt show`, we needed the ability to fetch only the rows that users specified via command-line flag. The behavior shipped in `1.5` would needlessly fetch the entire user-supplied query, then after return the specified number of rows.

The new argument is called `limit`
([source](https://github.com/dbt-labs/dbt-core/blob/ff5cb7ba51b4133f836d8d45ee8bb52f01ff4b4e/core/dbt/adapters/sql/connections.py#L143C72-L143C77))

#### How to support 

If your adapter over-rides SQLConnectionManager.execute(), you must include `limit` in it's function signature

Additionally, if your adapter overrides dbt-core's [`BaseAdapter.execute()`](https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/adapters/base/impl.py#L275-L290), you must also update that method to include the `limit` parameter.

#### What if you do nothing 

Things likely won't work for end users

</details>

<details>

<summary>[BtS] Adapter zone tests</summary>

The first step before starting to the upgrade process is to sure to bump the version of `dbt-tests-adapter`

```md
# after release cut
dbt-tests-adapter==1.6.0rc1
# after final release
dbt-tests-adapter~=1.6.0 
```

#### New tests

There are more tests in the adapter-zone test suite ([`tests/adapter/dbt/tests/adapter/`](https://github.com/dbt-labs/dbt-core/tree/main/tests/adapter/dbt/tests/adapter)). Some tests were introduced for new features and others to cover bugs that were fixed for this minor version

Within using the following command
```sh
git diff --unified=0 -G "class Test.*" v1.5.0...v1.6.0rc1 tests/adapter/dbt/tests/adapter | grep -E 'class Test.*'
```

below is a non-exhaustive list of some of the newly introduced tests

- `TestIncrementalConstraintsRollback`
- `TestTableContractSqlHeader`
- `TestIncrementalContractSqlHeader`
- `TestModelConstraintsRuntimeEnforcement`
- `TestConstraintQuotedColumn`
- `TestEquals`
- `TestMixedNullCompare`
- `TestNullCompare`
- `TestPostgresCloneNotPossible`
- `TestValidateSqlMethod`

</details>

**Comment by dataders** [comment]
using a comment to track progress across all (known) adapters

<details>

<summary>List 1.6 upgrade issues</summary>

- [x] https://github.com/databricks/dbt-databricks/issues/380
- [x] https://github.com/starburstdata/dbt-trino/issues/314
- [x] https://github.com/dbt-athena/dbt-athena/issues/347
- [x] https://github.com/jwills/dbt-duckdb/issues/202
- [x] https://github.com/aws-samples/dbt-glue/issues/191
- [x] https://github.com/microsoft/dbt-fabric/issues/41
- [x] https://github.com/oracle/dbt-oracle/issues/91
- [x] https://github.com/tglunde/dbt-exasol/issues/81
- [x] https://github.com/MaterializeInc/materialize/issues/20235
- [x] https://github.com/Teradata/dbt-teradata/issues/89
- [ ] https://github.com/ClickHouse/dbt-clickhouse/issues/164
- [ ] https://github.com/firebolt-db/dbt-firebolt/issues/109
- [ ] https://github.com/markporoshin/dbt-greenplum/issues/24
- [ ] https://github.com/cloudera/dbt-hive/issues/105
- [ ] https://github.com/aurany/dbt-ibmdb2/issues/38
- [ ] https://github.com/cloudera/dbt-impala/issues/176
- [ ] https://github.com/iomete/dbt-iomete/issues/7
- [ ] https://github.com/mindsdb/dbt-mindsdb/issues/30
- [ ] https://github.com/dbeatty10/dbt-mysql/issues/155
- [ ] https://github.com/chop-analytics/dbt-netezza/issues/20
- [ ] https://github.com/memsql/dbt-singlestore/issues/21
- [ ] https://github.com/splitgraph/sgr/issues/812
- [ ] https://github.com/dbt-msft/dbt-sqlserver/issues/421
- [ ] https://github.com/codeforkjeff/dbt-sqlite/issues/46
- [ ] https://github.com/microsoft/dbt-synapse/issues/179
- [ ] https://github.com/pingcap/dbt-tidb/issues/51
- [ ] https://github.com/vertica/dbt-vertica/issues/102
- [ ] https://github.com/dremio/dbt-dremio/issues/191
- [ ] https://github.com/fal-ai/fal/issues/880
- [ ] https://github.com/databendcloud/dbt-databend/issues/3
- [ ] https://github.com/inferlabs/dbt-infer/issues/8
- [ ] https://github.com/getindata/dbt-flink-adapter/issues/64
- [ ] https://github.com/tanyshak/dbt-upsolver/issues/4

</details>

**Comment by damian3031** [comment]
Hi @dataders, now that [1.6.0rc1](https://github.com/dbt-labs/dbt-core/releases/tag/v1.6.0rc1) has already been released, do you plan to update `Areas to consider` 🚧 sections of this guide?

**Reply by dataders** [reply]
hi @damian3031! I just updated them now. Let me know if you have any questions.

**Comment by okramarenko** [comment]
Hi everyone! Are you still planning to add info on how to stub materialized views elegantly? Thanks in advance :)
